### PR TITLE
[SSO] find organizational accounts by email – WEB-346

### PR DESCRIPTION
**Pull request by AlexTugarev**
First try to look up accounts by `[subjectID, audience]` as composite key, then try to lookup by email address. The scope of this lookup is limited to the Org owning the SSO configuration.

## Related Issue(s)

Fixes WEB-346

## How to test
* See unit test
* Regular SSO login should still work

#### Preview status

gitpod:summary

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold

### CODE REVIEW

File: components/gitpod-db/src/user-db.spec.db.ts [hunk 1]
```diff
@@ -176,6 +176,24 @@ class UserDBSpec {
...
+            authName: "Tester",
+            authProviderId: "oauth2-client-id",
+            primaryEmail: "tester@some.org",
+        });
+        orgUser = await this.db.storeUser(orgUser);
+
+        const result = await this.db.findOrganizationalUser("org1", "tester@some.org");
+        expect(result, "organizational user should be found").not.to.be.undefined;
+        expect(result!.identities, "should find a single identity").to.have.length(1);
+    }
```
Comment (easyCZ) [inline components/gitpod-db/src/user-db.spec.db.ts:196]: Would it be possible to add a negative record as well, one which we'll not find? 
So it should be in the same org, but the email shouldn't match

File: components/gitpod-db/src/typeorm/user-db-impl.ts [hunk 2]
```diff
@@ -599,6 +599,27 @@ export class TypeORMUserDBImpl implements UserDB {
             .getCount();
         return blockedUsers > 0;
     }
+
+    async findOrganizationalUser(organizationId: string, email: string): Promise<MaybeUser> {
```
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/user-db-impl.ts:603]: Would be great to find a common name for the "user who is owned by the Org". Sven used "owned" as the term here https://github.com/gitpod-io/gitpod/pull/17538/files#diff-4c72608b2e3eb106601a9bd44fbb13c2c551329302994c01dcab2586f68192d0R46 so in this case, I'd name it

```suggestion
    async findOwnedUsers(organizationId: string, email: string): Promise<MaybeUser> {
```

or 
```suggestion
    async getUser(organizationId: string, email: string, owned: bool): Promise<MaybeUser> {
```
here, the signature and types communicate the intent such that we don't need to push that into the method description.

2. Do we need to guarantee that the user is not only owned by the Org, but also that they have a relationship with the org through the TeamMembership table?

File: components/server/src/user/user-service.ts [hunk 3]
```diff
@@ -330,11 +330,16 @@ export class UserService {
...
     }
 
-    async findUserForLogin(params: { candidate: Identity }) {
+    async findUserForLogin(params: { candidate: IdentityLookup }) {
         let user = await this.userDb.findUserByIdentity(params.candidate);
         return user;
     }
 
+    async findOrganizationalUser(params: { organizationId: string; email: string }): Promise<MaybeUser> {
+        let user = await this.userDb.findOrganizationalUser(params.organizationId, params.email);
```
Comment (easyCZ) [inline components/server/src/user/user-service.ts:339]: Do we want some log-fields on this to be able to trace that later?

**Review by easyCZ — APPROVED** [review]
Functionally looks good

File: components/server/src/user/user-service.ts (hunk 3 above)
Comment (AlexTugarev) [reply components/server/src/user/user-service.ts]: done

File: components/gitpod-db/src/typeorm/user-db-impl.ts (hunk 2 above)
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/user-db-impl.ts]: renamed.

File: components/gitpod-db/src/user-db.spec.db.ts (hunk 1 above)
Comment (AlexTugarev) [reply components/gitpod-db/src/user-db.spec.db.ts]: done

**Pull request merged**